README.md
<p align="center">
    <img alt="Logo" src="assets/logo.svg" width="25%" />
</p>

<h1 align="center">
    Gradle Play Publisher
</h1>

<p align="center">
    <a href="https://travis-ci.org/Triple-T/gradle-play-publisher">
        <img src="https://img.shields.io/travis/Triple-T/gradle-play-publisher/master.svg?style=flat-square" />
    </a>
    <a href="https://search.maven.org/search?q=g:com.github.triplet.gradle%20AND%20a:play-publisher&core=gav">
        <img src="https://maven-badges.herokuapp.com/maven-central/com.github.triplet.gradle/play-publisher/badge.svg" />
    </a>
</p>

Gradle Play Publisher is a Gradle plugin that allows you to upload your App Bundle or APK and other
app details to the Google Play Store.

> **Note:** this README targets the `v2.0.0` betas. Documentation for the latest stable release is
available [here](https://github.com/Triple-T/gradle-play-publisher/blob/1.2.2/README.md).

## Table of contents

1. [Quickstart guide](#quickstart-guide)
1. [Prerequisites](#prerequisites)
   1. [Initial Play Store upload](#initial-play-store-upload)
   1. [Signing configuration](#signing-configuration)
   1. [Service Account](#service-account)
1. [Basic setup](#basic-setup)
   1. [Installation](#installation)
   1. [Authenticating Gradle Play Publisher](#authenticating-gradle-play-publisher)
1. [Task organization](#task-organization)
1. [Managing artifacts](#managing-artifacts)
   1. [Common configuration](#common-configuration)
   1. [Publishing an App Bundle](#publishing-an-app-bundle)
   1. [Publishing APKs](#publishing-apks)
   1. [Promoting artifacts](#promoting-artifacts)
   1. [Handling version conflicts](#handling-version-conflicts)
1. [Managing Play Store metadata](#managing-play-store-metadata)
   1. [Quickstart](#quickstart)
   1. [Directory structure](#directory-structure)
   1. [Publishing listings](#publishing-listings)
   1. [Publishing in-app products](#publishing-in-app-products)
1. [Advanced topics](#advanced-topics)
   1. [Using CLI options](#using-cli-options)
   1. [Encrypting Service Account keys](#encrypting-service-account-keys)
   1. [Using multiple Service Accounts](#using-multiple-service-accounts)

## Quickstart guide

1. Upload the first version of your APK or App Bundle using the
   [Google Play Console](https://play.google.com/apps/publish)
1. [Create a Google Play Service Account](#service-account)
1. [Sign your release builds](https://developer.android.com/studio/publish/app-signing#gradle-sign)
   with a valid `signingConfig`
1. [Add and apply the plugin](#installation)
1. [Authenticate GPP](#authenticating-gradle-play-publisher)

## Prerequisites

### Initial Play Store upload

The first APK or App Bundle needs to be uploaded via the Google Play Console because registering the
app with the Play Store cannot be done using the Play Developer API. For all subsequent uploads and
changes, GPP may be used.

### Signing configuration

If no publishing tasks were created, you most likely haven't added a valid signing configuration to
your release builds. Be sure to
[add one](https://developer.android.com/studio/publish/app-signing#gradle-sign).

### Service Account

To use GPP, you must create a service account with access to the Play Developer API. You'll find a
guide to setting up the service account
[here](https://developers.google.com/android-publisher/getting_started#using_a_service_account).
Once that's done, you'll need to grant at least the following permissions to your service account
for GPP to work (go to Settings -> Developer account -> Users & permissions):

<p>
    <img alt="Minimum Service Account permissions" src="assets/min-perms.png" width="66%" />
</p>

## Basic setup

### Installation

In your root `build.gradle(.kts)` file, add the Gradle Play Publisher dependency:

```kt
buildscript {
    repositories {
        // ...
        jcenter()
    }

    dependencies {
        // ...
        classpath("com.github.triplet.gradle:play-publisher:2.0.0-beta1")
    }
}
```

Then apply it to each individual `com.android.application` module where you want to use GPP.
For example, `app/build.gradle(.kts)` is a commonly used app module:

```groovy
apply plugin: 'com.android.application'
apply plugin: 'com.github.triplet.play'
```

#### Snapshot builds

If you're prepared to cut yourself on the bleeding edge of GPP development, snapshot builds are
available from
[Sonatype's `snapshots` repository](https://oss.sonatype.org/content/repositories/snapshots/):

```groovy
buildscript {
    repositories {
        // ...
        maven { url 'https://oss.sonatype.org/content/repositories/snapshots' }
    }

    dependencies {
        // ...
        classpath 'com.github.triplet.gradle:play-publisher:2.0.0-SNAPSHOT'
    }
}
```

### Authenticating Gradle Play Publisher

After you've gone through the [Service Account setup](#service-account), you should have a JSON file
with your private key. Add a `play` block alongside your `android` one with the file's location:

```kt
android { ... }

play {
    serviceAccountCredentials = file("your-key.json")
}
```

#### Using a PKCS12 key instead

```kt
play {
    serviceAccountEmail = "service-account-name@project-id.iam.gserviceaccount.com"
    serviceAccountCredentials = file("your-key.p12")
}
```

## Task organization

GPP follows the Android Gradle Plugin's naming convention: `[action][Variant][Thing]`. For example,
`publishPaidReleaseBundle` will be generated if have a `paid` product flavor.

Lifecycle tasks to publish multiple product flavors at once are also available. For example,
`publishBundle` publishes all variants.

To find available tasks, run `./gradlew tasks` and look under the publishing section.

## Managing artifacts

GPP supports uploading both the App Bundle and APK. Once uploaded, GPP also supports promoting those
artifacts.

### Common configuration

Several options are available to customize how your artifacts are published:

* `track` is the target stage for an artifact, i.e. alpha/beta/prod
* `releaseStatus` is the type of release, i.e. draft/completed/in progress
* `userFraction` is the percentage of users who will received a staged rollout

Example configuration:

```kt
play {
    // ...
    track = "rollout"
    userFraction = 0.5
    releaseStatus = "inProgress"
}
```

#### Uploading release notes

While GPP can automatically build and find your artifact, you'll need to tell the plugin about your
release notes.

Add a file under `src/[sourceSet]/play/release-notes/[language]/[track].txt` where `sourceSet`
is a [full variant name](https://developer.android.com/studio/build/build-variants#sourceset-build),
`language` is one of the
[Play Store supported codes](https://support.google.com/googleplay/android-developer/answer/3125566),
and `track` is the channel you want these release notes to apply to (or `default` if unspecified).

As an example, let's assume you have these two different release notes:

```
src/main/play/release-notes/en-US/default.txt
.../beta.txt
```

When you publish to the beta channel, the `beta.txt` release notes will be uploaded. For any other
channel, `default.txt` will be uploaded.

> Note: the Play Store limits your release notes to a maximum of 500 characters.

### Publishing an App Bundle

Run `./gradlew publishBundle`.

#### Defaulting to the App Bundle

You'll notice that if you run `./gradlew publish`, it uploads an APK by default. To change this,
default to the App Bundle:

```kt
play {
    // ...
    defaultToAppBundles = true
}
```

### Publishing APKs

Run `./gradlew publishApk`. Splits will be uploaded if available.

### Promoting artifacts

Existing releases can be promoted and/or updated to the [configured track](#common-configuration)
with `./gradlew promoteArtifact`.

By default, the track from which to promote a release will be determined by the most unstable
channel that contains a release. Example: if the alpha channel has no releases, but the beta and
prod channels do, the beta channel will be picked. To configure this manually, use the `fromTrack`
property:

```kt
play {
    // ...
    fromTrack = "alpha"
} 
```

### Handling version conflicts

If an artifact already exists with a version code greater than or equal to the one you're trying to
upload, an error will be thrown when attempting to publish the new artifact. You have two options:

* Ignore the error and continue (`ignore`)
* Automatically pick the correct version code so you don't have to manually update it (`auto`)

Example configuration:

```kt
play {
    // ...
    resolutionStrategy = "ignore"
}
```

#### Post-processing outputs sanitized by auto resolution

For example, you could update you app's version name based on the new version code:

```kt
play {
    // ...
    resolutionStrategy = "auto"
    outputProcessor = { // this: ApkVariantOutput
        versionNameOverride = "$versionNameOverride.$versionCode"
    }
}
```

## Managing Play Store metadata

GPP supports uploading any metadata you might want to change with each release, from screenshots and
descriptions to in-app purchases and subscriptions.

### Quickstart

GPP includes a bootstrap task that pulls down your existing listing and initializes everything for
you. To use it, run `./gradlew bootstrap`.

### Directory structure

GPP follows the Android Gradle Plugin's source set
[guidelines and priorities](https://developer.android.com/studio/build/build-variants#sourceset-build).
`src/[sourceSet]/play` is the base directory for Play Store metadata. Since `main` is the most
common source set, it will be assumed in all following examples.

In addition to merging metadata across variants, GPP merges translations. That is, if a resources is
provided in a default language such as `en-US` but not in `fr-FR`, the resource will be copied over
when uploading French metadata.

### Publishing listings

Run `./gradlew publishListing`.

#### Uploading global app metadata

Base directory: `play`

File | Description
--- | ---
`contact-email.txt` | Developer email
`contact-phone.txt` | Developer phone
`contact-website.txt` | Developer website
`default-language.txt` | The default language for both your Play Store listing and translation merging as described above

#### Uploading text based listings

Base directory: `play/listings/[language]` where `language` is one of the
[Play Store supported codes](https://support.google.com/googleplay/android-developer/answer/3125566)

File | Description | Character limit
--- | --- | ---
`title.txt`| App title | 50
`short-description.txt` | Tagline | 80
`full-description.txt` | Full description | 4000
`video-url.txt` | Youtube product video | N/A

#### Uploading graphic bases listings

Directory: `play/listings/[language]/graphics` where `language` is defined as in the previous
section

Image files are organized a bit differently than in previous sections. Instead of the file name, the
parent directory's name is used as the media type. This is because multiple images may be provided
for the same media type. While file names are arbitrary, they will be uploaded in alphabetical order
and presented on the Play Store as such. Therefore, we recommend using a number as the file name
(`1.png` for example). Both PNG and JPEG images are supported.

Directory | Max # of images
--- | ---
`icon` | 1
`feature-graphic` | 1
`promo-graphic` | 1
`phone-screenshots` | 8
`tablet-screenshots` | 8
`large-tablet-screenshots` | 8
`tv-banner` | 1
`tv-screenshots` | 8
`wear-screenshots` | 8

### Publishing in-app products

Run `./gradlew publishProducts`.

Manually setting up in-app purchase files is not recommended. [Bootstrap them instead](#quickstart)
with `./gradlew bootstrap --products`.

## Advanced topics

### Using CLI options

All configuration options available in the `play` block are also available as CLI options so you
don't have to update your build file when making one-time changes. For example, to configure
`play.track` on demand, use the `--track` option. `camelCase` options are converted to
`kebab-case` ones.

To get a list of options and their quick documentation, use `./gradlew help --task [task]` where
`task` is something like `publishBundle`.

### Encrypting Service Account keys

If you commit unencrypted Service Account keys to source, you run the risk of letting anyone access
your Google Play account. To circumvent this issue, many CI servers support encrypting files while
keeping fake versions in public source control. Here is a set of
[common fake files](https://github.com/SUPERCILEX/Robot-Scouter/tree/38407b3d6db74edb6c9de33b862655dfbd010a70/ci-dummies)
you might need and ways to encrypt your real keys for a few common CI servers:

- [Travis CI](https://docs.travis-ci.com/user/encrypting-files/)
- [CircleCI](https://github.com/circleci/encrypted-files)
- [Jenkins](https://github.com/samrocketman/jervis/wiki/Secure-secrets-in-repositories)

### Using multiple Service Accounts

If you need to publish each build flavor to a separate Play Store account, GPP supports flavor
specific `playAccountConfigs`:

```groovy
android {
    playAccountConfigs {
        firstCustomerAccount {
            serviceAccountCredentials = file('customer-one-key.json')
        }

        secondCustomerAccount {
            serviceAccountCredentials = file('customer-two-key.json')
        }
    }

    productFlavors {
        firstCustomer {
            playAccountConfig = playAccountConfigs.firstCustomerAccount
        }

        secondCustomer {
            playAccountConfig = playAccountConfigs.secondCustomerAccount
        }
    }
}
```


CHANGELOG.md
## 2.0.0 - to be released

### Features and improvements

* Support Android App Bundles - #262, #319
* Support promoting existing releases - #389, #284, #104
* Support publishing in-app purchases - #322, #181
* Allow overriding configuration via command line options - #326
* Allow bootstrapping only certain listing metadata - #375, #372
* Fully support multi-dimensional flavors - #130, #311
* Fill missing metadata with data from the default language - #107, #323
* Support Gradle cache and incremental builds - #304, #308
* Support Gradle task configuration avoidance - #388
* Support different resolution strategies in case of version conflicts - #301
* Add group tasks to publish all variants at once - #117, #273
* Move all public tasks to "Publishing" group instead of "Play Store" - #365
* Improve error messages and logging - #238, #268
* Add status indicators and upload progress - #298

### Bug fixes

* Don't upload the universal APK if splits are provided - #393, #380
* Provide publishing tasks even when `signingConfig` is missing - #244, #298
* Don't upload 0 byte mapping files - #370, #319
* A ton of other fixes we didn't track 😅

### Maintenance

* Convert plugin to Kotlin
* Bump Android Publisher plugin version to v3
* Moved package from `de.triplet.gradle.play` to `com.github.triplet.gradle.play`

### Breaking changes

#### Updated listings structure

As a precondition for supporting multi-dimensional flavors, the structure of listings metadata
has changed:

For example, english listing files have moved:

```
../play/en-US/listing/shortdescription --> ../play/listings/en-US/shortdescription
```

TODO: link to README with new structure, add link to migration tool

#### Removed configuration properties

* `uploadImages`: Since the plugin now makes use of Gradle's caching system and incremental builds
  (#308) to only upload images if they have changed, this property has become obsolete.
* `untrackOld`: With the introduction of conflict resolution strategies (#301), this property has
  become obsolete.
* `errorOnSizeLimit`: The plugin will now always error on size limit to provide deterministic
  behavior.

##### Simplified Service Account credentials API

The `jsonFile` and `pk12File` properties have been replaced with a unified
`serviceAccountCredentials` property.

#### Renamed tasks to follow AGP conventions

For example, `publishApkRelease` -> `publishReleaseApk`. Note: the old tasks are still available for
now, but will be removed in a future release.

#### AGP no longer bundled

The AGP needs to be depended on explicitly.

## 1.2.2 - 2018-05-24

* More descriptive error message when texts exceed allowed length - #172
* Update character limit for app title to 50 characters - #224, #229
* Support for internal test track - #253

## 1.2.1 - 2018-05-23

*Broken artifact on MavenCentral*

## 1.2.0 - 2017-03-23

* Support for multiple service accounts - #76, #161
* Support ABI splits - #39
* Support for uploading proguard mapping files - #132
* Automatically untrack older versions in other channels during upload - #121
* Make sure to only read the first line for items that expect a single line string - #143
* Remove trailing linebreaks - #187

## 1.1.5 - 2016-08-09

* Support Filipino in Metadata - #142
* Support Screenshots for Wear devices - #127
* Fixed character count for line breaks on Windows - #136

## 1.1.4 - 2015-10-14

* Do not require signingConfig for bootstrap and listing tasks - #100
* Support JSON file credentials - #97

## 1.1.3 - 2015-09-01

* Support staged rollouts - #79, #80
* Bugfix: Bootstrap into main source set for projects without flavors - #87

## 1.1.2 - 2015-07-16

* Bugfix: Corretly set contact web parameter - #82

## 1.1.1 - 2015-07-11

* Support for tv related images - #77
* Support for app details - #44
* Addes some explanation to google's applicationNotFound exception - #75

## 1.1.0 - 2015-05-09

* Breaking change: Renamed plugin to `com.github.triplet.play` so it does not collide with gradle's
  internal play plugin - #50
* Throw errors if texts exceed the size limits - #66

## 1.0.4 - 2015-04-14

* Bootstrap and upload promo video urls - #60

## 1.0.3 - 2015-04-13

* Defer retrieval of applicationId to task execution - #54
* Bugfix: Use UTF-8 encoding in bootstrap task - #59

## 1.0.2 - 2015-02-15

* Support build type specific resources - #38
* Bugfix: Trim texts for listing before upload - #41
* Skip upload of recent changes if play folder does not exist - #40
* Exclude guava-jdk5 dependency that conflicts with a newer version pulled in by the Android
  Plugin - #21

## 1.0.1 - 2015-02-06

* Create publish tasks for all non-debuggable build types
* Better error message for unsupported locales

## 1.0.0 - 2015-01-07

* Updated to Android plugin 1.0.0
* Find APK file based on the variant output. Eliminates workarounds for custom APK names
* Upload images in alphabetical order

## 0.14.1 - 2014-11-15

* Additional check if zipAlign task exists
* Use variant's applicationId instead of parsing the manifest

## 0.14.0 - 2014-11-12

* Updated to Android plugin 0.14.+
* Changed versioning scheme to better reflect what version of the Android plugin is supported
* Added descriptions to publisher tasks
* Added task `bootstrapReleasePlayResources` (and flavors) to help setting up the folder structure
* Allow separate `whatsnew` text files for different tracks

## 0.0.4 - 2014-10-11

* Added separate tasks for uploading the apk and for updating the listing
* Allow uploading screenshots and images with property `uploadImages true`


src/main/kotlin/de/triplet/gradle/play/Constants.kt
package de.triplet.gradle.play

import com.google.api.client.googleapis.javanet.GoogleNetHttpTransport
import com.google.api.client.http.AbstractInputStreamContent
import com.google.api.client.http.FileContent
import com.google.api.client.json.jackson2.JacksonFactory
import com.google.api.services.androidpublisher.model.Image
import java.io.File

internal const val APPLICATION_NAME = "gradle-play-publisher"
internal const val PLAY_STORE_GROUP = "Play Store"

internal const val MIME_TYPE_APK = "application/vnd.android.package-archive"
internal const val MIME_TYPE_IMAGE = "image/*"

internal const val LISTING_PATH = "listing/"
internal const val RESOURCES_OUTPUT_PATH = "build/outputs/play"

internal const val MAX_SCREENSHOTS_SIZE = 8

internal enum class ListingDetails(val fileName: String, val maxLength: Int = -1) {
    ContactEmail("contactEmail"),
    ContactPhone("contactPhone"),
    ContactWebsite("contactWebsite"),
    DefaultLanguage("defaultLanguage"),
    Title("title", 50),
    ShortDescription("shortdescription", 80),
    FullDescription("fulldescription", 4000),
    Video("video"),
    WhatsNew("whatsnew", 500);

    fun saveText(dir: File, value: String) = File(dir.apply { mkdirs() }, fileName).writeText(value)
}

internal val JSON_FACTORY by lazy { JacksonFactory.getDefaultInstance() }
internal val HTTP_TRANSPORT by lazy { GoogleNetHttpTransport.newTrustedTransport() }

internal val TRACKS = arrayOf("internal", "alpha", "beta", "rollout", "production")
internal val HIGHEST_TRACK = "internal"
internal val IMAGE_EXTENSIONS = arrayOf("png", "jpg")

internal enum class ImageTypes(val fileName: String, val max: Int = MAX_SCREENSHOTS_SIZE) {
    Icon("icon", 1),
    FeatureGraphic("featureGraphic", 1),
    PhoneScreenshots("phoneScreenshots"),
    SevenInchScreenshots("sevenInchScreenshots"),
    TenInchScreenshots("tenInchScreenshots"),
    PromoGraphic("promoGraphic", 1),
    TVBanner("tvBanner", 1),
    TVScreenshots("tvScreenshots"),
    WearScreenshots("wearScreenshots");

    fun getImages(listingDir: File): List<AbstractInputStreamContent>? {
        val graphicDir = File(listingDir, fileName)
        if (graphicDir.exists()) {
            return graphicDir.listFiles(ImageFileFilter())
                    .asList()
                    .sorted()
                    .map { FileContent(MIME_TYPE_IMAGE, it) }
        }
        return null
    }

    fun saveImages(listingDir: File, images: List<Image>?) {
        @Suppress("UNUSED_VARIABLE")
        val imageFolder = listingDir.validSubFolder(fileName) ?: return

        if (images == null) {
            return
        }

        // TODO: Disabled for now as we have only access to preview-versions with the current API.
        /*
        for (image in images) {
            File(imageFolder, "${image.id}.png").outputStream().use { os ->
                URL(image.url).openStream().use { it.copyTo(os) }
            }
        }
        */
    }


}

// region '419' is a special case in the play store that represents latin america
// 'fil' is a special case in the play store that represents Filipino
internal val LOCALE_REGEX = Regex("^(fil|[a-z]{2}(-([A-Z]{2}|419))?)\\z")
